mmlearn/datasets/processors/masking.py

```diff
@@ -264,3 +265,134 @@ def apply_masks(
 
     # Concatenate along the batch dimension
     return torch.cat(all_x, dim=0)
+
+
+@dataclass
+class IJEPAMaskGenerator:
+    """Generates encoder and predictor masks for preprocessing.
+
+    This class generates masks dynamically for individual examples and can be passed to
+    a data loader as a preprocessing step.
+
+    Parameters
+    ----------
+    input_size : tuple[int, int], default=(224, 224)
+        Input image size.
+    patch_size : int, default=16
+        Size of each patch.
+    min_keep : int, default=4
+        Minimum number of patches to keep.
+    allow_overlap : bool, default=False
+        Whether to allow overlap between encoder and predictor masks.
+    """
+
+    input_size: Tuple[int, int] = (224, 224)
+    patch_size: int = 16
+    min_keep: int = 4
+    allow_overlap: bool = False
+
+    def __post_init__(self) -> None:
+        """Initialize the mask generator."""
+        self.height = self.input_size[0] // self.patch_size
+        self.width = self.input_size[1] // self.patch_size
+
+    def _sample_block_size(
+        self,
+        generator: torch.Generator,
+        scale: Tuple[float, float],
+        aspect_ratio: Tuple[float, float],
+    ) -> Tuple[int, int]:
+        """Sample the size of the mask block based on scale and aspect ratio."""
+        _rand = torch.rand(1, generator=generator).item()
+        min_s, max_s = scale
+        mask_scale = min_s + _rand * (max_s - min_s)
+        max_keep = int(self.height * self.width * mask_scale)
+
+        min_ar, max_ar = aspect_ratio
+        aspect_ratio_val = min_ar + _rand * (max_ar - min_ar)
+
+        h = int(round(math.sqrt(max_keep * aspect_ratio_val)))
+        w = int(round(math.sqrt(max_keep / aspect_ratio_val)))
+
+        h = min(h, self.height - 1)
+        w = min(w, self.width - 1)
+
+        return h, w
+
+    def _sample_block_mask(
+        self,
+        b_size: Tuple[int, int],
+        acceptable_regions: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Sample a mask block."""
+        h, w = b_size
+        top = torch.randint(0, self.height - h, (1,)).item()
+        left = torch.randint(0, self.width - w, (1,)).item()
+        mask = torch.zeros((self.height, self.width), dtype=torch.int32)
+        mask[top : top + h, left : left + w] = 1
+
+        mask_complement = torch.ones((self.height, self.width), dtype=torch.int32)
+        mask_complement[top : top + h, left : left + w] = 0
+
+        return mask.flatten(), mask_complement.flatten()
+
+    def __call__(
+        self,
+        example: torch.Tensor,
+        enc_mask_scale: Tuple[float, float] = (0.2, 0.8),
+        pred_mask_scale: Tuple[float, float] = (0.2, 0.8),
+        aspect_ratio: Tuple[float, float] = (0.3, 3.0),
+        nenc: int = 1,
+        npred: int = 2,
```

Comment:
Do these change every batch? If not, I think we can define them as instance variables.